libnavui-androidauto/src/main/java/com/mapbox/androidauto/navigation/CarActiveGuidanceMarkers.kt

```diff
@@ -1,63 +1,54 @@
 package com.mapbox.androidauto.navigation
 
 import com.mapbox.androidauto.internal.extensions.getStyle
-import com.mapbox.androidauto.internal.extensions.handleStyleOnAttached
-import com.mapbox.androidauto.internal.extensions.handleStyleOnDetached
-import com.mapbox.androidauto.internal.extensions.mapboxNavigationForward
+import com.mapbox.androidauto.internal.extensions.styleFlow
 import com.mapbox.androidauto.internal.logAndroidAuto
 import com.mapbox.androidauto.placeslistonmap.PlacesListOnMapLayerUtil
+import com.mapbox.androidauto.routes.CarRoutesProvider
+import com.mapbox.androidauto.routes.NavigationCarRoutesProvider
 import com.mapbox.geojson.Feature
 import com.mapbox.geojson.FeatureCollection
+import com.mapbox.maps.Style
 import com.mapbox.maps.extension.androidauto.MapboxCarMapObserver
 import com.mapbox.maps.extension.androidauto.MapboxCarMapSurface
-import com.mapbox.maps.plugin.delegates.listeners.OnStyleLoadedListener
-import com.mapbox.navigation.core.MapboxNavigation
-import com.mapbox.navigation.core.directions.session.RoutesObserver
-import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.base.route.NavigationRoute
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.launch
+
+class CarActiveGuidanceMarkers(
+    private val carRoutesProvider: CarRoutesProvider = NavigationCarRoutesProvider(),
+) : MapboxCarMapObserver {
 
-class CarActiveGuidanceMarkers : MapboxCarMapObserver {
-    private var styleLoadedListener: OnStyleLoadedListener? = null
-    private var mapboxCarMapSurface: MapboxCarMapSurface? = null
-    private val navigationObserver = mapboxNavigationForward(this::onAttached, this::onDetached)
-    private val routesObserver = RoutesObserver { updateMarkers() }
     private val placesLayerUtil = PlacesListOnMapLayerUtil()
+    private lateinit var scope: CoroutineScope
 
-    private fun updateMarkers() {
-        val route = MapboxNavigationApp.current()?.getNavigationRoutes()?.firstOrNull()
-            ?: return
-        val coordinate = route.routeOptions.coordinatesList().lastOrNull()
-            ?: return
-        val featureCollection = FeatureCollection.fromFeature(Feature.fromGeometry(coordinate))
-        mapboxCarMapSurface?.getStyle()?.let {
-            placesLayerUtil.updatePlacesListOnMapLayer(it, featureCollection)
-        }
+    private fun updateMarkers(style: Style, routes: List<NavigationRoute>) {
+        val features = routes.take(1)
```

Comment:
I like yours but also was thinking about alternatives

```
    private fun updateMarkers(style: Style, routes: List<NavigationRoute>) {
        val destination = routes.firstOrNull { it.routeOptions.coordinatesList().isNotEmpty() }
            ?.let { listOf(Feature.fromGeometry(it.routeOptions.coordinatesList().last())) }
            ?: emptyList<Feature>()
        placesLayerUtil.updatePlacesListOnMapLayer(style, FeatureCollection.fromFeatures(destination))
    }
```

Reply:
Yours is a bit different from what we had before.

Reply:
is it? this is updating a destination marker right? I didn't test it i just considered logic the same

Reply:
previously the null values would return. now it always calls `updatePlacesListOnMapLayer` so it seems there may have been an issue fixed 🤔

Reply:
You are right, mine is also different from what we had before. I think it is better to explicitly clear the markers in this case.